### PR TITLE
Cheat Engine

### PR DESCRIPTION
**Pull request by RipleyTom**
First draft of a cheat manager.

Note that the cheats you make probably won't survive another launch of the game if the offset is on the heap. I may add a more complex parameter to make some level of redirection possible in the future.

edit:
I added a redirection system, completely untested, syntax is:
$ is offset
[] for dereferencing
+/- for adding offset
ex: [[$]+8]+16
It will get the u32 offset is pointing at, add 8 to that, get the u32 that is pointing to and add 16 to that which will be the final offset.

If the Script cell is left blank it will act as before.

### CODE REVIEW

File: rpcs3/rpcs3qt/main_window.cpp [hunk 1]
```diff
@@ -1331,6 +1332,11 @@ void main_window::CreateConnects()
 		trop_manager->show();
 	});
 
+	connect(ui->actionManage_Cheats, &QAction::triggered, [=] {
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/main_window.cpp]: newline. (I'm starting to like it like this though... work, you know XD)

File: rpcs3/rpcs3qt/cheat_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+	lst_search->setContextMenuPolicy(Qt::CustomContextMenu);
+	grp_add_cheat_layout->addWidget(lst_search);
+	grp_add_cheat->setLayout(grp_add_cheat_layout);
+	main_layout->addWidget(grp_add_cheat);
+
+	setLayout(main_layout);
+
+	// Edit/Manage UI
+	connect(tbl_cheats, &QTableWidget::itemClicked, [=](QTableWidgetItem* item) {
+		int row           = tbl_cheats->currentRow();
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/cheat_manager.cpp]: row can be -1 if you click on empty space

File: rpcs3/rpcs3qt/cheat_manager.cpp [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+			edt_value_final->setText(cur_value);
+		}
+		else
+		{
+			log_cheat.fatal("Failed to retrieve cheat selected from internal cheat_engine");
+		}
+	});
+
+	connect(tbl_cheats, &QTableWidget::cellChanged, [=](int row, int column) {
+		if (column != 1)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/cheat_manager.cpp]: also check for row maybe

File: rpcs3/rpcs3qt/cheat_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,660 @@
...
+	case signed_64_cheat:
+	{
+		s64 to_search = qstr_to_search.toLongLong(&res_conv);
+		if (res_conv)
+			offsets_found = g_cheat.search(to_search, offsets_found);
+		break;
+	}
+	}
+
+	if (!res_conv)
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/cheat_manager.cpp]:
```suggestion
	if (res_conv)
	{
		offsets_found = g_cheat.search(to_search, offsets_found);
	}
	else
```

File: rpcs3/rpcs3qt/cheat_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,660 @@
...
+	connect(btn_new_search, &QPushButton::clicked, [=](int action) {
+		offsets_found.clear();
+		do_the_search();
+	});
+
+	connect(btn_filter_results, &QPushButton::clicked, [=](int action) { do_the_search(); });
+
+	connect(lst_search, &QListWidget::customContextMenuRequested, [=](const QPoint& loc) {
+		QPoint globalPos      = lst_search->mapToGlobal(loc);
+		QMenu* menu           = new QMenu();
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/cheat_manager.cpp]: put menu below return

File: rpcs3/rpcs3qt/cheat_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,660 @@
...
+				break;
+			}
+			case signed_64_cheat:
+			{
+				s64 to_set = qstr_to_apply.toLongLong(&res_conv);
+				if (res_conv)
+					res_apply = g_cheat.set_value(cheat->offset, to_set);
+				break;
+			}
+			}
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/cheat_manager.cpp]: default missing

File: rpcs3/rpcs3qt/cheat_manager.cpp (hunk 2 above)
Comment (RipleyTom) [reply rpcs3/rpcs3qt/cheat_manager.cpp]: I put a check just in case.

File: rpcs3/rpcs3qt/cheat_manager.cpp (hunk 3 above)
Comment (RipleyTom) [reply rpcs3/rpcs3qt/cheat_manager.cpp]: This is only called when a cell has been edited(after the edit) so I think it's ok like that?

File: rpcs3/rpcs3qt/cheat_manager.cpp (hunk 4 above)
Comment (RipleyTom) [reply rpcs3/rpcs3qt/cheat_manager.cpp]: I can't do that because search is a template and I need to specify or let the compiler deduce the template type.

**Comment by RipleyTom** [comment]
Fixed a bug where saves wouldn't be loaded properly if script was empty.
Cleaned up the code(no more duplicate code, const functions are now const and static functions are now static).
I tried to implement a function to try to find a pointer to address back in "safe" memory(data segment) but from testing it doesn't seem to work very well so I disabled it but left the code in, in case anyone comes up with an improvement.

File: rpcs3/rpcs3qt/cheat_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,859 @@
...
+	be_t<T> value_swapped = value;
+
+	if (Emu.IsStopped())
+		return {};
+
+	bool did_we_pause = false;
+	if (Emu.IsRunning())
+	{
+		did_we_pause = true;
+		Emu.Pause();
```
Comment (elad335) [inline rpcs3/rpcs3qt/cheat_manager.cpp]: This is unsafe, use `cpu_thread::suspend_all cpu_lock(nullptr);` to suspend execution of all threads. (scope based)

Comment (elad335) [reply rpcs3/rpcs3qt/cheat_manager.cpp]: It works regardless if the emu is paused or not.

File: rpcs3/rpcs3qt/cheat_manager.cpp [hunk 8]
```diff
@@ -0,0 +1,840 @@
...
+					results.push_back(off);
+			}
+		}
+	}
+	else
+	{
+		// Looks through mapped memory
+		for (u32 page_ind = (0x10000 / 4096); page_ind < (0xF0000000 / 4096); page_ind++)
+		{
+			if (vm::check_addr(page_ind * 4096, 1, vm::page_info_t::page_allocated | vm::page_info_t::page_readable))
```
Comment (elad335) [inline rpcs3/rpcs3qt/cheat_manager.cpp]: I know you are aware of this, leaving it here as a reminder.
```suggestion
			if (vm::check_addr(page_ind * 4096))
```

File: rpcs3/rpcs3qt/cheat_manager.cpp [hunk 9]
```diff
@@ -0,0 +1,840 @@
...
+	if (Emu.IsStopped())
+		return {};
+
+	cpu_thread::suspend_all cpu_lock(nullptr);
+
+	if (to_filter.size())
+	{
+		for (const auto& off : to_filter)
+		{
+			if (vm::check_addr(off, 1, vm::page_info_t::page_allocated | vm::page_info_t::page_readable))
```
Comment (elad335) [inline rpcs3/rpcs3qt/cheat_manager.cpp]: Here as well.

File: rpcs3/rpcs3qt/cheat_manager.cpp [hunk 10]
```diff
@@ -0,0 +1,840 @@
...
+
+	T ret_value = *vm::get_super_ptr<T>(offset);
+
+	return ret_value;
+}
+
+template <typename T>
+bool cheat_engine::set_value(const u32 offset, const T value)
+{
+	if (!vm::check_addr(offset, vm::page_info_t::page_allocated | vm::page_info_t::page_writable))
```
Comment (elad335) [inline rpcs3/rpcs3qt/cheat_manager.cpp]: Missing a cpu_lock here?

Comment (RipleyTom) [reply rpcs3/rpcs3qt/cheat_manager.cpp]: Good catch!

File: rpcs3/rpcs3qt/cheat_manager.cpp [hunk 11]
```diff
@@ -0,0 +1,860 @@
...
+			success = false;
+
+		result = static_cast<T>(result_16);
+	}
+
+	if constexpr (std::is_same<T, u16>::value)
+		result = str.toUShort(&success);
+
+	if constexpr (std::is_same<T, u32>::value)
+		result = str.toULong(&success);
```
Comment (Nekotekina) [inline rpcs3/rpcs3qt/cheat_manager.cpp]: toUInt

File: rpcs3/rpcs3qt/cheat_manager.cpp [hunk 12]
```diff
@@ -0,0 +1,860 @@
...
+		if (result_16 < -128 || result_16 > 127)
+			success = false;
+
+		result = static_cast<T>(result_16);
+	}
+
+	if constexpr (std::is_same<T, s16>::value)
+		result = str.toShort(&success);
+
+	if constexpr (std::is_same<T, s32>::value)
```
Comment (Nekotekina) [inline rpcs3/rpcs3qt/cheat_manager.cpp]: toInt

**Pull request merged**